### PR TITLE
Advanced Mute Engine

### PR DESCRIPTION
**Pull request by pixeldesu**
This feature adds the ability to extend BetterTweetDeck with custom filters easily. 🎉 

Before this, each custom filter would require taking `TD.vo.Filter.prototype.pass` apart and adding a new switch. 

To not break _working_ behaviour of TweetDeck, the Advanced Mute Engine (AME) includes fallbacks to the old functions used.

The AME has following features:
* Add custom filters
* Extends `getDisplayType` automatically, so the filter list looks just fine
* Extends the global filter list template with custom filters that should be displayed in the dropdown  
  _You can hide filters from the dropdown (e.g. if you want to use them somewhere else)_

The AME implementation is documented throughout, to not cause confusion on what it's actually doing. I also included documentation on the custom filters, how to add them etc. (see `docs/extend/filters.md`)

This PR also includes the first custom filter (Hiding Retweets from Users) as a reference implementation.

---

I went sure that regular BTD would not break with the code submitted in this PR. To surely mitigate any upcoming issues it's recommended to let this code get through an extended beta-testing period.

### CODE REVIEW

**Comment by pixeldesu** [comment]
This PR is ready 🎉 

If there's anything I need to change, leave a review etc. and I'll adjust accordingly.

File: docs/extend/filters.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+Mute Filters are used by TweetDeck to hide unwanted content. BetterTweetDeck adds an advanced mute engine which enables developers/contributors to
+add their own filters to the client.
+
+## How Filters work in TweetDeck
+
+Every displayable object in TweetDeck runs through every filter that is set active globally or for specific columns. In order that a tweet/interaction
+is shown, it needs to `pass` filters. If a tweet passes, it's shown. Simple as that.
+
+So, in the function of your custom filter, you need to return `false` (or `t.positive` as an alternative) if something matches your filter. If your filter
+returns `true` it's assumed that this tweet doesn't match the current filter and is shown.
```
Comment (eramdam) [inline docs/extend/filters.md]: I feel like a warning somewhere about performances and the fact that the function must be synchronous could be useful.

Comment (pixeldesu) [reply docs/extend/filters.md]: added!

**Comment by pixeldesu** [comment]
**Note:** This feature will solve #253

**Pull request merged**